crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2423,6 +2464,56 @@ impl From<bool> for Truthiness {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Len {
+    /// Can be inferred statically.
+    Known(u64),
+    /// Cannot be inferred statically, falling back to `int`.
+    Unknown,
```

Comment:
Minor point: maybe rename this to `StaticallyUnknown` or similar to avoid any confusion with `Type::Unknown`? (especially since you map `Len::Never` to `Type::Never`)